src/Common/Session/SessionUtil.php

```diff
@@ -143,6 +143,26 @@ public static function portalSessionCookieDestroy(): void
         self::standardSessionCookieDestroy();
     }
 
+    public static function oauthSessionStart(): void
+    {
+        session_start([
+            'cookie_samesite' => "Strict",
```

Comment:
This will work on localhost but fail on different origin and reinit session. e.g login form and getting authserver saved session back in authorize token.

Reply:
I actually initially set this to `None`, but then noted that it tested well on `Strict`. `Lax` would also be an option. We can definitely change this, if needed (my brain takes a while to work out csrf type stuff like this; I couldn't find places in code that were accepting form submissions from an outside server, but again, my brain has a tough time with this stuff).

Reply:
In reality, login should be an iframe and setup to communicate with a distant iframe or client via messaging channel or session states. Also and though I don't allow currently, client can dictate through request attributes whether or not to display a login and/or initiate a login session. Same with terminating and clean exiting and on and on;)
In the end, origin (client) submits the form. This is why I chose to stay away from ajax/es6 and doing validations in php.

So the state of the session was being destroyed by openemrs session manager between start of auth and user consent where, like you pointed out, SameSite is most likely the issue.

As said, all works well from localhost but fails when origin is outside of openemr. It's best to test all of flows from a secure domain into openemr secure domain.

I allow http for now but, should go soon.

Reply:
Are you able to test this setting (None) on a separate origin to ensure this is needed (with the code in this PR). Note that the current session setup in master did not work for me either (the session was being killed every time login was called which agree probably related to some weird clashing with openemr's session), so there may of been other reasons why you were having session issues.

Reply:
As an aside, it would be nice to have a non-gui way to connect back with auth credentials (ie. the app can collect it on its own and then send it back in a api call) after getting the code grant type request (note i am still very naive on oauth processes, so maybe this is not possible).

Reply:
I pull this branch down and test.


> As an aside, it would be nice to have a non-gui w

I'm not done with this forms workflow yet. It needs to be a dialog and once I get persisting authentication finished, we'll have the hooks we need for other interruptions in flow.

Reply:
lol, man O man, you don't throw anything away on your repo do ya?

Reply:
gotta maintain a historical archive :)

Reply:
soon I'll have to use random uuid's for my branch names to ensure no conflicts :)

Reply:
That's why we're glad you're around! Still it'd drive me nutzzz.
Build is looking very nice!

Reply:
Most all pathing is broken. We need fully qualified urls and beside having to add to global.inc we have to get from server!
We need protocol, domain and port is any. e.g discovery

  |  
-- | --
issuer | ""
authorization_endpoint | "/oauth2/default/authorize"
token_endpoint | "/oauth2/default/token"
jwks_uri | "/oauth2/default/jwk"
userinfo_endpoint | "/oauth2/default/userinfo"
registration_endpoint | "/oauth2/default/registration"

Using webroot is precarious at best. Normally in core we can get away with such because browser will build out the full url. Not so here.
For containers and sitting in environments like proxies etc i'm unsure. Seem you and me have been here before with portal where we rely on browser as work around. We may have to rely on a preset base uri from our new global.

Reply:
It is working on my end. Am I missing something or not testing correctly (basically making calls from postman except for the login endpoint get which am running directly in browser to get the login redirect)? (ie. why do we need full urls?)
(i am pretty sure getting a full url from server(ie. apache or nginx) will be impossible in many contexts)

Reply:
And if needed, can we use:
```
$GLOBALS['site_addr_oath'] . $GLOBALS['webroot'] . stuff already there
```

Reply:
oh, I think i know what you are talking about. that code can be fixed by using:
```
$base_url = $authServer->authBaseFullUrl;
```
(i'll push fix to this pr soon)

Reply:
just pushed a fix for this

Reply:
Throughout the entire auth flow the client does one thing first, completely pulls in the discovery endpoint json so it can get endpoints for the rest of the flow. Then at certain points in the flow namely one being claims, client needs the id/access tokens issuer to verify claims iss attribute. This is done by comparing the claims iss to the discovery issuer which is the base uri of protocol, domain etc. This is another security point of transaction.
This is not the only reason for discovery. We could publish all our api endpoints and thus restrict if needed.

So you see, this reason alone requires a full url. Also, i'm not so sure not having on server side is not going to have issues.
Testing on localhost doesn't reveal the entire picture.

Reply:
looks like i have more fun stuff to learn :)
will get the client you emailed me to work so i can test better and don't break this stuff

Reply:
Okay Brady, i'm now officially befuddled and turned around on these urls. I get we need to keep this straight for other server contexts and you're on the right track. I just need to hit pause for a moment so I can wrap my head around the shift from my design intent.
Right now I have issues with encryption key decrypt from the one I created yesterday. I deleted to get by but concerning.
Also seeing session issues so I need to explain those. 
We could have kept my url structure and just added the new global for domain.

Reply:
To reset encryption key and key pair, just delete the private key (if that is missing then it will rebuild the key pair and encryption key and replace lingering public key and encryption key if they exist).

Sounds like issuer needs to just be that site global (without the port and webroot) which is something I can accomplish easily and will push another pr soon to address that.

Goal is to end up with same url structure that you had without asking apache/nginx for it (since that won't work a lot of the time); we are almost there when get the issuer cleared up :) 

Will push the quick PR for the issuer fix. After that will spend time getting the oauth2 client up and working so I don't break things on other planned updates (and also confirm pathing is working).

Reply:
> Goal is to end up with same url structure that you had without asking apache/nginx for it

Yep, I see that and is fine. Just let me catch up with testing before ya do anymore commits to urls. I think what you have will be okay.

Reply:
> After that will spend time getting the oauth2 client up and working

That client will help with flow but unless you run from different domain as one you're testing then won't fully be a correct test.

Reply:
sure thing. will hold off on url work and will work on getting client up and testing :)

Reply:
Quick update. Firstly the urls are fine the way they are currently. Discovery by client works well sooo yeah, a small victory.
Full circle to where I started, sessions. Turns out a combination of issues are causing problems. cookie_samesite needs to be None and cookie_secure = true. Also a fyi, you can't use session name to switch between sessions. This is why I originally used session_id(). Session name is the cookie name in our use.
 Still, either way we still get into trouble with use_strict_mode which we don't want to loosen that requirement so i'm going to need to alter workflow or find a way around session not being recalled on token issuance endpoint.